File: structure_d/validation/retry.py
"""Retry / fallback logic for validation failures."""

from __future__ import annotations

import json
from typing import Type

import structlog
from pydantic import BaseModel

from structure_d.config import get_settings
from structure_d.exceptions import InferenceError
from structure_d.inference.providers import BaseLLMProvider, VLLMProvider
from structure_d.inference.structured_output import StructuredOutputBuilder
from structure_d.schemas.base import ExtractionResult, TaskType
from structure_d.validation.validator import SchemaValidator

logger = structlog.get_logger(__name__)


class RetryHandler:
    """
    Validate extraction results and retry on failure.

    Flow
    ----
    1. If the result is already valid (``is_valid=True``), return immediately.
       Providers that use constrained decoding or native structured outputs
       guarantee validity, so re-validation is redundant.
    2. If invalid, call ``provider.generate()`` with a *refined prompt* that
       includes the original text and the error descriptions, giving the model
       a chance to self-correct.
    3. Repeat up to ``max_retries`` times.

    The :class:`~structure_d.validation.validator.SchemaValidator` is kept as
    a utility for callers that need raw-string validation outside this flow.
    """

    def __init__(
        self,
        schema_cls: Type[BaseModel],
        task: TaskType = TaskType.EXTRACTION,
        provider: BaseLLMProvider | None = None,
        max_retries: int | None = None,
    ) -> None:
        settings = get_settings()
        self.schema_cls = schema_cls
        self.task = task
        self.validator = SchemaValidator(schema_cls)
        self.provider = provider or VLLMProvider()
        self.max_retries = max_retries or settings.validation.max_retries
        self.retry_with_prompt = settings.validation.retry_with_refined_prompt
        self.builder = StructuredOutputBuilder(schema_cls=schema_cls, task=task)

    async def validate_and_retry(
        self,
        result: ExtractionResult,
        original_text: str,
        model: str | None = None,
    ) -> ExtractionResult:
        """
        Ensure *result* is valid, retrying via the provider if it is not.

        Mutates and returns the same :class:`ExtractionResult` instance.
        """
        # Fast path: provider already validated the output.
        if result.is_valid:
            return result

        errors = result.validation_errors or ["Provider returned invalid or empty output"]

        if not self.retry_with_prompt:
            result.validation_errors = errors
            return result

        for attempt in range(1, self.max_retries + 1):
            logger.info(
                "validation_retry",
                attempt=attempt,
                errors=errors,
                chunk_id=result.chunk_id,
            )

            refined_prompt = self._build_refined_user_prompt(original_text, errors)
            system_prompt = self.builder.build_system_prompt()

            try:
                pr = await self.provider.generate(
                    prompt=refined_prompt,
                    schema=self.schema_cls,
                    system_prompt=system_prompt,
                    model=model,
                )
                result.raw_output = pr.raw_text or json.dumps(pr.output.model_dump())
                result.structured_output = pr.output.model_dump()
                result.model_used = pr.model_used
                result.is_valid = True
                result.validation_errors = []
                return result
            except InferenceError as exc:
                errors = [str(exc)]
                logger.warning(
                    "validation_retry_failed",
                    attempt=attempt,
                    error=str(exc),
                    chunk_id=result.chunk_id,
                )

        result.validation_errors = errors
        return result

    # ── Helpers ───────────────────────────────────────────────────────────────

    @staticmethod
    def _build_refined_user_prompt(original_text: str, errors: list[str]) -> str:
        """Build a retry user prompt that includes prior validation errors."""
        error_block = "\n".join(f"- {e}" for e in errors)
        return (
            f"The previous extraction had the following validation errors:\n"
            f"{error_block}\n\n"
            f"Please re-extract the data from the text below, fixing all errors.\n\n"
            f"{original_text}"
        )

File: cli/src/validation/retry.rs
use serde_json::Value;

use crate::inference::{GenerateRequest, LLMProvider, ProviderResult};
use super::{validate, ValidationResult};

pub struct RetryHandler {
    pub max_retries: usize,
}

impl RetryHandler {
    pub fn new(max_retries: usize) -> Self {
        Self { max_retries }
    }

    pub async fn validate_and_retry(
        &self,
        result: &ProviderResult,
        original_text: &str,
        schema: Option<&Value>,
        provider: &dyn LLMProvider,
        temperature: f32,
        max_tokens: usize,
    ) -> ValidationResult {
        let mut validation = validate(&result.content);

        if validation.is_valid {
            return validation;
        }

        // Retry with refined prompt
        for attempt in 0..self.max_retries {
            let refined_prompt = build_refined_prompt(original_text, &validation.errors, attempt);

            let mut req = GenerateRequest::new(&refined_prompt)
                .with_temperature(temperature)
                .with_max_tokens(max_tokens);

            if let Some(s) = schema {
                req = req.with_schema(s);
            }

            match provider.generate(req).await {
                Ok(retry_result) => {
                    validation = validate(&retry_result.content);
                    if validation.is_valid {
                        tracing::debug!(
                            attempt = attempt + 1,
                            "Validation succeeded on retry"
                        );
                        return validation;
                    }
                }
                Err(e) => {
                    tracing::warn!(error = %e, attempt = attempt + 1, "Retry attempt failed");
                    break;
                }
            }
        }

        tracing::warn!(
            retries = self.max_retries,
            "All retry attempts exhausted, returning invalid result"
        );
        validation
    }
}

fn build_refined_prompt(original_text: &str, errors: &[String], attempt: usize) -> String {
    let error_list = errors
        .iter()
        .enumerate()
        .map(|(i, e)| format!("  {}. {}", i + 1, e))
        .collect::<Vec<_>>()
        .join("\n");

    format!(
        "Your previous response could not be parsed as valid JSON (attempt {}).\n\
         Errors:\n{}\n\n\
         Please extract the structured data from the following text and return ONLY \
         valid JSON, no explanation, no markdown fences:\n\n{}",
        attempt + 1,
        error_list,
        original_text
    )
}
